src/Lean/Elab/Tactic/BVAckermannize.lean

```diff
@@ -0,0 +1,447 @@
+    return ({ defn, eqProof}, g)
+
+/-- Insert the CallInfo `cv` at `(fn, args)` into the state. -/
+private def _insertCallVal (fn : Function) (args : ArgumentArray) (cv : CallInfo) : AckM Unit := do
+  let calls ← _getCallMap fn
+  modify fun s => { s with fn2args2call := s.fn2args2call.insert fn (calls.insert args cv) }
+
+/--
+Replace a call `f x₁ ... xₙ` with a new free variable `fv` where `fv := f x₁ ... xₙ`.
+This free variable `fv` is cached, and represented by a `CallVal`.
+Moreover, Since the `fv` is defeq to `f x₁ ... xₙ, we can substitute `fv` for `f x₁ ... xₙ`.
+-/
+def insertOrLookupCall (g : MVarId) (fn : Function) (args : ArgumentArray) : AckM (CallInfo × MVarId) := do
+  g.withContext do
+    if let some val ← getCallVal? fn args then
+      trace[Meta.Tactic.bv_ack] "using cached call {val} for {fn} {args}"
+      return (val, g)
+    let e := mkAppN fn.f (args.map Argument.value)
+    let name ← gensym
+    let (introDef, g) ← introDefExt g name fn.codomain.toExpr e
+    let cv := { fvar := introDef.defn, heqProof := Expr.fvar introDef.eqProof : CallInfo }
+    _insertCallVal fn args cv
+    return (cv, g)
+
+/-- Create a trace node in trace class (i.e. `set_option traceClass true`),
+with header `header`, whose default collapsed state is `collapsed`. -/
+def withTraceNode (g : MVarId) (header : MessageData) (k : AckM α)
+    (collapsed : Bool := true)
+    (traceClass : Name := `Meta.Tactic.bv_ack) : AckM α :=
+  withContext g do Lean.withTraceNode traceClass (fun _ => return header) k (collapsed := collapsed)
+
+/--
+The result of running ackermannization on an expression. Returns the ackermannized expression,
+as well as informing whether this expression can be used as a subexpression for further ackermannization,
+by tracking whether the expression has loose bound variables.
+-/
+structure AckResult where
+  /-- The resulting expression from ackermannization. -/
+  val : Expr
+  /--
+  Whether an expression has bound variables in it.
+  If it does, then it cannot be used as the subexpression of an ackermanized call,
+  since the expression is not zeroth-order.
+  -/
+  hasLooseBvar : Bool
+deriving Inhabited
+
+
+/--
+Given a binary effectful computation `f : α → β → m γ`, and two arrays `as : Array α` and `bs : Array β`,
+build an array whose elements are created by invoking `f as[i] bs[i]` in sequence.
+This obeys the equation `zipWithM as bs f = as.zip bs |>.mapM f
+-/
+private def _root_.Array.zipWithM [Monad m] (as : Array α) (bs : Array β) (f : α → β → m γ) : m (Array γ) := do
+  let mut out := #[]
+  for (a, b) in as.zip bs do
+    out := out.push (← f a b)
+  return out
+
+mutual
+
+/--
+Try to ackermannize an application. Recursively ackermannize all of its subexpressions,
+and try to ackermannize the application if:
+- Neither the function `f` nor the arguments `x₁` ... `xₙ` have bound variables in them.
+- `f` has a type that is ackermannizable: all arguments and output are either `BitVec w` or `Bool`.
+-/
```

Comment:
```suggestion
Try to ackermannize an application. 
First, recursively ackermannize all of its subexpressions,
then try to ackermannize the application itself if:
- Neither the function `f` nor the arguments `x₁` ... `xₙ` have bound variables in them.
- `f` has a type that is ackermannizable: all arguments and output are either `BitVec w` or `Bool`.
-/
```

I assume this is what's meant? The original phrasing confused me a bit, since I read it as: first ackermannize the application, then ackermannize the subexpressions, then ackermannize the application (huh?).